src/react/types.ts

```diff
@@ -1,30 +1,18 @@
-import { CSSProperties, ReactNode } from "react";
-
 export type ViewportComponentAttributes = Pick<
   React.HTMLAttributes<HTMLElement>,
   "className" | "style" | "id" | "role" | "tabIndex" | "onKeyDown" | "onWheel"
 > &
   React.AriaAttributes;
 
-export interface CustomContainerComponentProps {
-  style: CSSProperties;
-  children: ReactNode;
-}
+export type CustomContainerComponentProps<R extends React.ElementType = 'div'> = React.ComponentPropsWithRef<R>
 
-export type CustomContainerComponent = React.ForwardRefExoticComponent<
-  React.PropsWithoutRef<CustomContainerComponentProps> &
-    React.RefAttributes<any>
->;
+export type CustomContainerComponent<R extends React.ElementType = 'div'> = React.FunctionComponent<CustomContainerComponentProps<R>>;
 
 /**
  * Props of customized item component for {@link Virtualizer} or {@link WindowVirtualizer}.
  */
-export interface CustomItemComponentProps {
-  style: CSSProperties;
+export type CustomItemComponentProps<R extends React.ElementType = 'div'> = React.ComponentPropsWithRef<R> & {
   index: number;
-  children: ReactNode;
-}
+};
 
-export type CustomItemComponent = React.ForwardRefExoticComponent<
-  React.PropsWithoutRef<CustomItemComponentProps> & React.RefAttributes<any>
->;
+export type CustomItemComponent<R extends React.ElementType = 'div'> = React.FunctionComponent<CustomItemComponentProps<R>>;
```

Comment:
@aeharding 
Hi, how about changing the type like this? (CustomItemComponentProps is also reverted).

```suggestion
export type CustomItemComponent =
  | React.ForwardRefExoticComponent<
      CustomItemComponentProps & React.RefAttributes<any>
    >
  | FunctionComponent<CustomItemComponentPropsWithRef<any>>;

export interface CustomItemComponentPropsWithRef<T extends HTMLElement>
  extends CustomItemComponentProps {
  ref: RefObject<T>;
}
```

Your suggestion seems not typesafe because all fields are optional and it has fields that do not passed actually. I'm not sure this is perfect but it will solve the problems.